### ISSUE TITLE
Figure out an approach for adding experiments to the leaderboard

### ISSUE THREAD

**Issue by KennethEnevoldsen** [embeddings-benchmark/mteb#1211]
We currently do not have a good approach to adding experiment runs to the MTEB leaderboard e.g. experiment with the influence of hyperparameters. Such as using no-prompt for prompt-based models or changing the embedding size.

A solution to implement is to run it with a unique model_name, but there is currently no documentation on how one would need to do that.

---

edit: a potentially better solution is to add experiments. Where we add a layer to the visualization:

```
model_name
| - revision1
|    | - task1.json
     | - ...
     | - experiments
     |    | - exp_name1
     |    |    | -  task1.json
     |    |    | - ...
     |    | - exp_name1
     |    |    | -  task1.json
     |    |    | - ...
```

or potentially a slightly more consistent structure:

```
model_name
| - revision1
     | - default
     |    | -  task1.json
     |    | - ...
     | - exp_name1
     |    | -  task1.json
     |    | - ...
```

Experiment names could be e.g. no-instruct/instruct (whether you use instructions) or emb_size=256

To run a model as an experiment we could do:

```
model = MyModel(**custom_kwargs)
results = mteb.MTEB(model, experiment_name = "no-instruct")
```

**Comment by KennethEnevoldsen** [comment]
This way also allows us to filter experiments on the leaderboard.

**Comment by orionw** [comment]
I like this, gives more flexibility and cleans up the main folder.

The only potentially hard part is what do we do say for GritLM but for each type of prompt it runs?  The user would have to identify some title for the prompt to put down (e.g. E5 prompt or Instructor prompt or "prompt-I-just-made-up").  These will then be non-standard since every user will name it differently (or perhaps run different prompts).

**Comment by Samoed** [comment]
I think there could be a file for each experiment that includes the parameters used for the experiment, similar to `model_meta` or additional kwargs for model can be logged

**Comment by KennethEnevoldsen** [comment]
> The only potentially hard part is what do we do say for GritLM but for each type of prompt it runs? The user would have to identify some title for the prompt to put down (e.g. E5 prompt or Instructor prompt or "prompt-I-just-made-up"). These will then be non-standard since every user will name it differently (or perhaps run different prompts).

Def. agree with this. I would imagine that some standards form, but there will be some custom stuff to allow for flexibility.

> I think there could be a file for each experiment that includes the parameters used for the experiment, similar to model_meta. This can be done by logging additional kwargs for model

I think we can just log this in the model_meta?

**Comment by ayush1298** [comment]
Hey, Can anyone gives me an idea about status of this issue? I see the PR #1779 regarding tracking of kwargs. So, do we have to restructure/rename the results repo based on it?

**Comment by KennethEnevoldsen** [comment]
This is still not done. It requires more conceptual work, especially around how to display it in the leaderboards meaningfully. 

A solution could be something like (tracking only embedding size):

| model name           | embedding dim | results ... |
|----------------------|---------------|-------------|
| text-embedding-large | 1024          |  ... |
| text-embedding-large | 512           |  ... |

And figure out how to display this meaningfully on the leaderboard.

**Comment by Samoed** [comment]
And what if someone want to change prompt, normalization etc?

**Comment by KennethEnevoldsen** [comment]
> @KennethEnevoldsen embedding dim are already present in Summary table, So, Do you want to display these separately on leaderboard?

Yeah, it is, but it is not clear that it is a run setting (rather than just a static attribute of the model)

@Samoed: No, I completely agree, but if we can't even figure out the simple case, then it is hard to add the complex cases.

A simple solution is to move it here:

<img width="1285" alt="Image" src="https://github.com/user-attachments/assets/0bd75a06-0d55-452c-9ccd-62e507b660d4" />

Then, we can test out experiments where we only add "embedding-size" difference to the leaderboard (for now). 

WDYT?

**Comment by Samoed** [comment]
We can create directories as you've suggested in description and save metadata in `model_metadata.json` and show this additional data in leaderboard, but this would require duplication of model

**Comment by KennethEnevoldsen** [comment]
Yep. I don't think we can do that without duplicating the model

**Comment by ayush1298** [comment]
As #4071 is added, so I think only part which was missing here, is to add on LB. Can we do it in this way like, in summary table, we can have results for default settings (Need to define this one exactly) and add an extra tab called experiments and there we can add results with different kwrags of model. So, this won't confused people, as having all things in summary table will make it bigger and increase confusion

**Comment by Samoed** [comment]
I think we can add models with experiment params in name `model_name (param1=val1)`, but this could be hard to read

**Comment by ayush1298** [comment]
> I think we can add models with experiment params in name `model_name (param1=val1)`, but this could be hard to read

in params, we can have `embed_dim,` maybe `prompts,` or even if some special setting is applied, so how we could add all those in `model_name`

**Comment by KennethEnevoldsen** [comment]
I think `model_name (param1=val1)` is a good way to go and then we probably want to limit the type of experiments that we allow on the results repo to e.g. "embed-dim", "output-dtype" and maybe a few more (but that would be a great start)

**Comment by ayush1298** [comment]
> I think `model_name (param1=val1)` is a good way to go and then we probably want to limit the type of experiments that we allow on the results repo to e.g. "embed-dim", "output-dtype" and maybe a few more (but that would be a great start)

I thought of working on these and going through the thread once again. I agree with @Samoed that won't it will make difficult for people to read when we add model_name with parameters

**Comment by KennethEnevoldsen** [comment]
but then we need an alternative - we already have some models with (kwargs=value)

**Comment by Samoed** [comment]
> we already have some models with (kwargs=value)

I think this will be fine, because voyage uses name like this for experiments

**Comment by ayush1298** [comment]
okay, will take these then and implement it

**Comment by Samoed** [comment]
I don't think we want to add it now. We don't have such results in results repo, for now

**Comment by ayush1298** [comment]
> I don't think we want to add it now. We don't have such results in results repo, for now

We can make an issue to run them, and could run them

**Comment by Samoed** [comment]
> issue to run them

What do you want to run?

**Comment by ayush1298** [comment]
Running models with different kwargs and add their result to results repo, so they can be also shown on LB

**Comment by ayush1298** [comment]
Should I take it? Seems we have some results for experiments, though we have not discussed on adding prompts

**Comment by Samoed** [comment]
I don't think so. I'm refactoring leaderboard in https://github.com/embeddings-benchmark/mteb/pull/4716 and we want to completely refactor it soon

**Comment by ayush1298** [comment]
> I don't think so. I'm refactoring leaderboard in [#4716](https://github.com/embeddings-benchmark/mteb/pull/4716) and we want to completely refactor it soon

Isn't #4716 is only about speedup?

**Comment by Samoed** [comment]
It's about speed, but I refactor every part of leaderboard, so at least you need to wait for it

**Comment by ayush1298** [comment]
> It's about speed, but I refactor every part of leaderboard, so at least you need to wait for it

okay, will wait
